File: 15-3Sum.cpp
/*

Given an array S of n integers, are there elements a, b, c in S such that a + b + c = 0? Find all unique triplets in the array which gives the sum of zero.

Note: The solution set must not contain duplicate triplets.

For example, given array S = [-1, 0, 1, 2, -1, -4],

A solution set is:
[
  [-1, 0, 1],
  [-1, -1, 2]
]

*/


// http://www.geeksforgeeks.org/count-triplets-with-sum-smaller-that-a-given-value/

// http://www.geeksforgeeks.org/write-a-c-program-that-given-a-set-a-of-n-numbers-and-another-number-x-determines-whether-or-not-there-exist-two-elements-in-s-whose-sum-is-exactly-x/


// Time:  O(n^2)
// Space: O(1)

class Solution {

public:

    vector<vector<int>> threeSum(vector<int> &nums) {

        vector<vector<int>> res;
        const int target = 0;

        sort(nums.begin(), nums.end());

        for (int i = 0; i < static_cast<int>(nums.size()) - 2; ++i) {
            if (i == 0 || nums[i] != nums[i - 1]) {  // Skip duplicate.
                for (int j = i + 1, k = nums.size() - 1; j < k; ) {  // Time: O(n) for each i.
                    if (j  > i+1  && nums[j] == nums[j - 1]) {  // Skip duplicate.
                        ++j;
                    } 
                    else if (k + 1 < nums.size() && nums[k] == nums[k + 1]) {  // Skip duplicate.
                        --k;
                    } 
                    else {
                        const auto sum = nums[i] + nums[j] + nums[k];
                        if (sum > target) {  // Should decrease sum.
                            --k;
                        } 
                        else if (sum < target) {  // Should increase sum.
                            ++j;
                        } 
                        else {
                            res.push_back({nums[i], nums[j], nums[k]});
                            ++j, --k;
                        }
                    }
                }
            }
        }
        return res;
    }
};




class Solution {

public:

    vector<vector<int>> threeSum(vector<int> &nums) {

        vector<vector<int>> res;
        const int target = 0;

        sort(nums.begin(), nums.end());

        for (int i = 0; i < static_cast<int>(nums.size()) - 2; ++i) {
            if (i > 0 && nums[i] == nums[i - 1]) continue;  // Skip duplicate.
            for (int j = i + 1, k = nums.size() - 1; j < k; ) {  // Time: O(n) for each i.
                if (j  > i+1  && nums[j] == nums[j - 1]) {  // Skip duplicate.
                    ++j;
                } 
                else if (k < nums.size()-1 && nums[k] == nums[k + 1]) {  // Skip duplicate.
                        --k;
                } 
                else {
                    const auto sum = nums[i] + nums[j] + nums[k];
                    if (sum > target) {  // Should decrease sum.
                        --k;
                    } 
                    else if (sum < target) {  // Should increase sum.
                        ++j;
                    } 
                    else {
                        res.push_back({nums[i], nums[j], nums[k]});
                        ++j, --k;
                    }
                }
            }
        }
        return res;
    }
};


// Brute Force/ Naive O(n^3) solution

class Solution2 {

public:

    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> res;

        for (int i=0; i < nums.size(); ++i) {
            if(i > 0 && (nums[i] == nums[i-1])) continue;
            for (int j = i+1; j < nums.size(); ++j) {
                if(j > i+1 && (nums[j] == nums[j-1])) continue;
                for(int k = j+1; k < nums.size(); k++) {
                    if (k > j+1 && (nums[k] == nums[k-1])) continue;
                    if(nums[i] + nums[j] + nums[k] == 0) res.push_back({nums[i], nums[j], nums[k]});
                }
            }
        }

        return res;
    }
};

File: java/15-3Sum.java
/*

Given an array S of n integers, are there elements a, b, c in S such that a + b + c = 0? Find all unique triplets in the array which gives the sum of zero.

Note: The solution set must not contain duplicate triplets.

For example, given array S = [-1, 0, 1, 2, -1, -4],

A solution set is:
[
  [-1, 0, 1],
  [-1, -1, 2]
]

*/

public class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> res = new LinkedList<>();
        Arrays.sort(nums);
        int n = nums.length;
        int target = 0;
        for(int i = 0; i< n-2; i++) {
            if(i>0 && (nums[i] == nums[i-1])) continue;
            int j = i+1;
            int k = n-1;
            while(j < k) {
                if(j>i+1 && nums[j] == nums[j-1]) ++j;
                else if(k<n-1 && nums[k] == nums[k+1]) --k;
                else {
                    int sum = nums[i] + nums[j] + nums[k];
                    if( sum < target) ++j;
                    else if (sum > target) --k;
                    else {
                        List<Integer> l = new LinkedList<Integer>();
                        l.add(nums[i]);
                        l.add(nums[j]);
                        l.add(nums[k]);
                        res.add(l);
                        ++j;
                        --k;
                    }
                }
            }
        }
        return res;
    }
}
